### PR TITLE
Remove global single instance of `skaffoldOpts` in skaffold error package and use runContext Config

### PR DESCRIPTION
**Pull request by tejal29**
Review Order:
1) #5532 - event refactor to use Config instead of variables from config

When fixing #5485 , @briandealwis  [suggested](https://github.com/GoogleContainerTools/skaffold/pull/5491#pullrequestreview-606602375) removing the global `skaffoldOpts` in error package [here](https://github.com/GoogleContainerTools/skaffold/blob/c1b3f76b4668033c006d7905ab713ae6ea973278/pkg/skaffold/errors/errors.go#L46) 

In this PR
1)  Move `initErrors`, `buildErrors` and `deployErrors` to individual packages. 
2) Create `Problem` and allow packages to register problems.  The suggestions, will switch the `interface` to respective `Config` and suggest suggestions.
```
- type suggestionFunc func(runCtx *runcontext.RunContext) []*proto.Suggestion
+ type suggestionFunc func(cfg interface{}) []*proto.Suggestion
```
3) Change definition of `Suggestions` function to run on any interface and remove singleton use of `runCtx` or `skaffoldOpt` as mentioned by @briandealwis in his review feedback. 
So the following methods have changed
```
-func ShowAIError(err error) error {
+ func ShowAIError(cfg interface{}, err error) error {
```
and 
```
-func ActionableErr(phase Phase, err error) *proto.ActionableErr {
+func ActionableErr(cfg interface{}, phase Phase, err error) *proto.ActionableErr {
```

3) `ShowAIError` is now called in `cmd.withRunner` flow where `runCtx` is available and `handleWellknownErrors` is deleted. 
creatNewRunner also returns a `RunContext` to pass to error handling.
```
- func createNewRunner(out io.Writer, opts config.SkaffoldOptions) 
-    (runner.Runner, []*latest.SkaffoldConfig, error) {
+func createNewRunner(out io.Writer, opts config.SkaffoldOptions) 
   (runner.Runner, *runcontext.RunContext, []*latest.SkaffoldConfig, error) {
```

3) `ActionableErr` was called in `event` packages which used global singelton `skaffoldOpts`. We know pass `runCtx` during `event.InitializeState`
```
-func InitializeState(c []latest.Pipeline, kc string, autoBuild, autoDeploy, autoSync bool) {
+func InitializeState(cfg Config)
```
which is stored in `eventHandler`
```
type eventHandler struct {
	eventLog []proto.LogEntry
	logLock  sync.Mutex
+	cfg       Config
	state     proto.State
	stateLock sync.Mutex
	eventChan chan firedEvent
	listeners []*listener
}
```

Mutliple goroutines accessing `handler.cfg` should not cause a data race because this value is not only written once in `InitializeState`. 

See https://notes.shichao.io/gopl/ch9/#avoid-writing-the-variable

4) The calls to `ActionableErr` now use this to pass the `event.Config`.

### CODE REVIEW

File: pkg/skaffold/errors/deploy_problems.go [hunk 1]
```diff
@@ -29,20 +28,20 @@ var (
 	isMinikube = cluster.GetClient().IsMinikube
 )
 
-func suggestDeployFailedAction(runCtx runcontext.RunContext) []*proto.Suggestion {
-	if isMinikube(runCtx.KubeContext) {
+func suggestDeployFailedAction(cfg Config) []*proto.Suggestion {
+	if isMinikube(cfg.GetKubeContext()) {
```
Comment (briandealwis) [inline pkg/skaffold/errors/deploy_problems.go:32]: I'm fine with keeping this for now, but IMHO the config should have the _cluster_, which we can query if it's local, or specifically minikube or kind, etc.

And then we can have the clusters provide a status-check command.  The generic one is `kubectl describe nodes` and `kubectl get events`.

File: cmd/skaffold/app/cmd/runner.go [hunk 2]
```diff
@@ -46,32 +46,30 @@ import (
...
-	runner, config, err := createRunner(out, opts)
+	runner, runCtx, config, err := createRunner(out, opts)
 	if err != nil {
 		return err
 	}
 
 	err = action(runner, config)
 
-	return alwaysSucceedWhenCancelled(ctx, err)
+	return alwaysSucceedWhenCancelled(ctx, runCtx, err)
```
Comment (briandealwis) [inline cmd/skaffold/app/cmd/runner.go:56]: It's odd, but I don't see this change here?

File: cmd/skaffold/app/cmd/runner.go [hunk 3]
```diff
@@ -46,32 +46,30 @@ import (
...
 }
 
 // createNewRunner creates a Runner and returns the SkaffoldConfig associated with it.
-func createNewRunner(out io.Writer, opts config.SkaffoldOptions) (runner.Runner, []*latest.SkaffoldConfig, error) {
+func createNewRunner(out io.Writer, opts config.SkaffoldOptions) (runner.Runner, *runcontext.RunContext, []*latest.SkaffoldConfig, error) {
```
Comment (briandealwis) [inline cmd/skaffold/app/cmd/runner.go:60]: Does the runContext need to be explicitly returned, or could the runner.Runner have a `Context()` method?

**Review by briandealwis — COMMENTED** [review]
The `command` &rarr; `cmd` is odd?

**Comment by tejal29** [comment]
> The `command` → `cmd` is odd?

yeah maybe a search replace did. i just noticed it.

File: cmd/skaffold/app/cmd/runner.go (hunk 3 above)
Comment (tejal29) [reply cmd/skaffold/app/cmd/runner.go]: I added a `ErrConfig` to keep scope limited.

File: pkg/skaffold/errors/deploy_problems.go (hunk 1 above)
Comment (tejal29) [reply pkg/skaffold/errors/deploy_problems.go:32]: i think `config.ClusterType()` could be a good addition so we don't have to repeat the logic many times.

File: pkg/skaffold/errors/build_problems.go [hunk 4]
```diff
@@ -47,11 +46,11 @@ func suggestBuildPushAccessDeniedAction(runCtx runcontext.RunContext) []*proto.S
 	}
 
 	// check if global repo is set
-	if cfg, err := getConfigForCurrentContext(runCtx.Opts.GlobalConfig); err == nil {
+	if cfg, err := getConfigForCurrentContext(cfg.GlobalConfig()); err == nil {
 		if defaultRepo := cfg.DefaultRepo; defaultRepo != "" {
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/errors/build_problems.go:50]: um no. this code trying to distinguish between cases where
1)  the default repo is from command line `--default-repo`
2)  or via global configuration. 
let me rename the `cfg` with `gCfg` to indicate that

File: pkg/skaffold/docker/client.go [hunk 5]
```diff
@@ -56,7 +56,10 @@ type Config interface {
 	GetKubeContext() string
 	MinikubeProfile() string
 	GetInsecureRegistries() map[string]bool
+	Command() string
 	Mode() config.RunMode
+	DefaultRepo() *string
+	GlobalConfig() string
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/docker/client.go:62]: I agree. i should re-design this by not bleeding error config methods everywhere.

**Comment by tejal29** [comment]
@briandealwis hold off until i try again.

File: cmd/skaffold/app/cmd/cmd.go [hunk 6]
```diff
@@ -265,12 +266,14 @@ func setUpLogs(stdErr io.Writer, level string, timestamp bool) error {
...
 }
 
-func alwaysSucceedWhenCancelled(ctx context.Context, err error) error {
+func alwaysSucceedWhenCancelled(ctx context.Context, cfg sErrors.Config, err error) error {
 	// if the context was cancelled act as if all is well
-	if err != nil && ctx.Err() == context.Canceled {
+	if ctx.Err() == context.Canceled {
 		return nil
+	} else if err == context.Canceled {
+		return err
```
[the remark this answers is not in the document]
Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go:286]: added some comments and a TODO

**Comment by tejal29** [comment]
@briandealwis This is ready  for review again!

File: pkg/skaffold/build/build_problems_test.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+		description string
+		context     config.ContextConfig
+		optRepo     string
+		err         error
+		expected    string
+		expectedAE  *proto.ActionableErr
+	}{
+		{
+			description: "Push access denied when neither default repo or global config is defined",
+			err:         fmt.Errorf("skaffold build failed: could not push image: denied: push access to resource"),
```
Comment (briandealwis) [inline pkg/skaffold/build/build_problems_test.go:56]: I think all of these `fmt.Errorf()` can be replaced with just `errors.New()`

File: pkg/skaffold/build/build_problems.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+const (
+	PushImageErr = "could not push image"
+	// Error Prefix matches error thrown by Docker
+	// See https://github.com/moby/moby/blob/master/client/errors.go#L18
+	dockerConnectionFailed = ".*(Cannot connect to the Docker daemon.*) Is"
+
+	// Build cancelled due to other build failures
+	buildCancelled = ".*globalRepo canceled.*"
```
Comment (briandealwis) [inline pkg/skaffold/build/build_problems.go:36]:
```suggestion
	// See https://github.com/moby/moby/blob/master/client/errors.go#L20
	// `docker build: error during connect: Post \"https://127.0.0.1:32770/v1.24/build?buildargs=:  globalRepo canceled`
	dockerBuildCancelled = ".*globalRepo canceled.*"
```

File: pkg/skaffold/deploy/deploy_problems.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/GoogleContainerTools/skaffold/proto/v1"
+)
+
+func suggestDeployFailedAction(cfg interface{}) []*proto.Suggestion {
+	deployCfg, ok := cfg.(types.Config)
+	if !ok {
+		return nil
+	}
+	kCtx := deployCfg.GetKubeContext()
+	isMinikube := deployCfg.MinikubeProfile() != ""
```
Comment (briandealwis) [inline pkg/skaffold/deploy/deploy_problems.go:35]:
```suggestion
	isMinikube := cluster.GetClient().IsMinikube(kCtx)

```

File: pkg/skaffold/deploy/deploy_problems.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func init() {
+	sErrors.AddPhaseProblems(constants.Deploy, []sErrors.Problem{
+		{
+			Regexp:  re("(?i).*unable to connect.*: Get (.*)"),
+			ErrCode: proto.StatusCode_DEPLOY_CLUSTER_CONNECTION_ERR,
+			Description: func(err error) string {
+				matchExp := re("(?i).*unable to connect.*Get (.*)")
+				if match := matchExp.FindStringSubmatch(fmt.Sprintf("%s", err)); len(match) >= 2 {
```
Comment (briandealwis) [inline pkg/skaffold/deploy/deploy_problems.go:65]:
```suggestion
				if match := matchExp.FindStringSubmatch(err.Error()); len(match) >= 2 {
```

File: pkg/skaffold/errors/problem.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+func (p Problem) Error() string {
+	description := fmt.Sprintf("%s\n", p.Err)
+	if p.Description != nil {
+		description = p.Description(p.Err)
+	}
+	return description
+}
+
+func (p Problem) AIError(i interface{}, err error) error {
```
Comment (briandealwis) [inline pkg/skaffold/errors/problem.go:62]: This should have a doc comment.

File: pkg/skaffold/errors/problem.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+	p.Err = err
+	if p.Suggestion == nil {
+		return p
+	}
+	if suggestions := p.Suggestion(i); suggestions != nil {
```
Comment (briandealwis) [inline pkg/skaffold/errors/problem.go:67]:
```suggestion
	if suggestions := p.Suggestion(i); len(suggestions) > 0 {
```

File: pkg/skaffold/errors/problem.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+func isProblem(err error) (Problem, bool) {
+	if p, ok := err.(Problem); ok {
+		return p, true
+	}
+	return Problem{}, false
+}
+
+func AddPhaseProblems(phase constants.Phase, problems []Problem) {
+	if ps, ok := allErrors[phase]; ok {
+		allErrors[phase] = append(ps, problems...)
```
Comment (briandealwis) [inline pkg/skaffold/errors/problem.go:82]: Existing problems will be overwritten on line 84.
```suggestion
		problems = append(ps, problems...)
```

File: pkg/skaffold/initializer/init_problems.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+func init() {
+	sErrors.AddPhaseProblems(constants.Init, []sErrors.Problem{
+		{
+			Regexp:     re(".*creating tagger.*"),
+			ErrCode:    proto.StatusCode_INIT_CREATE_TAGGER_ERROR,
+			Suggestion: sErrors.ReportIssueSuggestion,
+		},
+		{
+			Regexp:      re(".*The control plane node must be running for this command.*"),
+			ErrCode:     proto.StatusCode_INIT_MINIKUBE_NOT_RUNNING_ERROR,
```
Comment (briandealwis) [inline pkg/skaffold/initializer/init_problems.go:41]: Isn't it possible we can see this error during deploy too?

File: pkg/skaffold/errors/errors.go [hunk 15]
```diff
@@ -18,45 +18,34 @@ package errors
...
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/instrumentation"
-	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner/runcontext"
 	"github.com/GoogleContainerTools/skaffold/proto/v1"
 	protoV2 "github.com/GoogleContainerTools/skaffold/proto/v2"
 )
 
 const (
 	// Report issue text
 	reportIssueText = "If above error is unexpected, please open an issue " + constants.GithubIssueLink + " to report this error"
```
Comment (briandealwis) [inline pkg/skaffold/errors/errors.go:31]: The links are quite big to fit in the middle of a sentence.
```suggestion
	reportIssueText = "If above error is unexpected, please open an issue to report this error at " " + constants.GithubIssueLink
```

File: pkg/skaffold/errors/errors.go [hunk 16]
```diff
@@ -79,54 +68,42 @@ func ActionableErrV2(phase constants.Phase, err error) *protoV2.ActionableErr {
...
 		for _, v := range problems {
-			if v.regexp.MatchString(err.Error()) {
-				return v.errCode, v.suggestion(runCtx)
+			if v.Regexp.MatchString(err.Error()) {
+				return v.ErrCode, v.Suggestion(cfg)
 			}
 		}
 	}
-	return proto.StatusCode_UNKNOWN_ERROR, nil
+	return unknownErrForPhase(phase), ReportIssueSuggestion(cfg)
```
Comment (briandealwis) [inline pkg/skaffold/errors/errors.go:106]: ❤️  the `ReportIssueSuggestion` here

File: pkg/skaffold/errors/errors.go [hunk 17]
```diff
@@ -79,54 +68,42 @@ func ActionableErrV2(phase constants.Phase, err error) *protoV2.ActionableErr {
...
 
-func getErrorCodeFromError(phase constants.Phase, err error) (proto.StatusCode, []*proto.Suggestion) {
+func getErrorCodeFromError(cfg interface{}, phase constants.Phase, err error) (proto.StatusCode, []*proto.Suggestion) {
 	var sErr Error
 	if errors.As(err, &sErr) {
 		return sErr.StatusCode(), sErr.Suggestions()
 	}
 
 	if problems, ok := allErrors[phase]; ok {
 		for _, v := range problems {
```
Comment (briandealwis) [inline pkg/skaffold/errors/errors.go:100]: Use `p` instead of `v`?

**Review by briandealwis — APPROVED** [review]
A bunch of suggestions below, but this is a great improvement.

File: cmd/skaffold/app/cmd/cmd.go [hunk 18]
```diff
@@ -268,12 +270,22 @@ func setUpLogs(stdErr io.Writer, level string, timestamp bool) error {
 	return nil
 }
 
-func alwaysSucceedWhenCancelled(ctx context.Context, err error) error {
+// alwaysSucceedWhenCancelled returns nil if the context was cancelled.
+// if the error is due to cancellation, plumbs it through and it gets swallowed
+// in skaffold main.
+// For all other errors, pass it through known errors.
+// TODO: Return null is error is of context.Cancelled and remove check in main.
```
[the remark this answers is not in the document]
Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go:277]: yeah. But the tests fail :(. We actually have tests which check the error return is cancelled.

File: cmd/skaffold/app/cmd/cmd.go [hunk 19]
```diff
@@ -268,12 +270,22 @@ func setUpLogs(stdErr io.Writer, level string, timestamp bool) error {
...
+func alwaysSucceedWhenCancelled(ctx context.Context, runCtx *runcontext.RunContext, err error) error {
+	if err == nil {
+		return err
+	}
 	// if the context was cancelled act as if all is well
-	if err != nil && ctx.Err() == context.Canceled {
+	if ctx.Err() == context.Canceled {
 		return nil
+	} else if err == context.Canceled {
```
[the remark this answers is not in the document]
Comment (tejal29) [reply cmd/skaffold/app/cmd/cmd.go:285]: I am not sure. There are tests in place for this and hence i added a TODO.

File: cmd/skaffold/app/cmd/runner.go [hunk 20]
```diff
@@ -53,16 +55,16 @@ func withRunner(ctx context.Context, out io.Writer, action func(runner.Runner, [
...
 	err = action(runner, config)
 
-	return alwaysSucceedWhenCancelled(ctx, err)
+	return alwaysSucceedWhenCancelled(ctx, runCtxErr, err)
 }
 
 // createNewRunner creates a Runner and returns the SkaffoldConfig associated with it.
 func createNewRunner(out io.Writer, opts config.SkaffoldOptions) (runner.Runner, []*latest.SkaffoldConfig, error) {
 	runCtx, configs, err := runContext(out, opts)
+	runCtxErr = runCtx
```
[the remark this answers is not in the document]
Comment (tejal29) [reply cmd/skaffold/app/cmd/runner.go:64]: You caught me being a little lazy here. Would need a lot of changes. Will do.

File: pkg/skaffold/initializer/init_problems.go (hunk 14 above)
Comment (tejal29) [reply pkg/skaffold/initializer/init_problems.go:41]: I think we do. @etanshaul opened an issue for this. Will fix it as part of that effort.

File: pkg/skaffold/build/build_problems.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+			Description: func(error) string {
+				return "Build Failed. No push access to specified image repository"
+			},
+			Suggestion: suggestBuildPushAccessDeniedAction,
+		},
+		{
+			Regexp:  re(buildCancelled),
+			ErrCode: proto.StatusCode_BUILD_CANCELLED,
+			Description: func(error) string {
+				return "Build Cancelled."
```
[the remark this answers is not in the document]
Comment (tejal29) [reply pkg/skaffold/build/build_problems.go:65]: They get trimmed anyways when concatenating. Will remove from here.

**Pull request merged**